File: bindings/zig/src/safe_math.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeMath - FFI bindings to libproven arithmetic operations.
// All computation is performed in verified Idris 2 code via libproven.

const c = @cImport(@cInclude("proven.h"));

/// Error types for math operations, mapped from ProvenStatus codes.
pub const MathError = error{
    DivisionByZero,
    Overflow,
    Underflow,
    InvalidArgument,
};

/// Map ProvenStatus to MathError.
fn mapStatus(status: i32) MathError!void {
    return switch (status) {
        c.PROVEN_OK => {},
        c.PROVEN_ERR_DIVISION_BY_ZERO => error.DivisionByZero,
        c.PROVEN_ERR_OVERFLOW => error.Overflow,
        c.PROVEN_ERR_UNDERFLOW => error.Underflow,
        else => error.InvalidArgument,
    };
}

/// Safe integer division via libproven.
/// Returns PROVEN_ERR_DIVISION_BY_ZERO if denominator is 0,
/// PROVEN_ERR_OVERFLOW for INT64_MIN / -1.
pub fn div(numerator: i64, denominator: i64) MathError!i64 {
    const result = c.proven_math_div(numerator, denominator);
    try mapStatus(result.status);
    return result.value;
}

/// Safe modulo operation via libproven.
/// Returns error if denominator is 0.
pub fn mod(numerator: i64, denominator: i64) MathError!i64 {
    const result = c.proven_math_mod(numerator, denominator);
    try mapStatus(result.status);
    return result.value;
}

/// Checked addition with overflow detection via libproven.
pub fn addChecked(a: i64, b: i64) MathError!i64 {
    const result = c.proven_math_add_checked(a, b);
    try mapStatus(result.status);
    return result.value;
}

/// Checked subtraction with underflow detection via libproven.
pub fn subChecked(a: i64, b: i64) MathError!i64 {
    const result = c.proven_math_sub_checked(a, b);
    try mapStatus(result.status);
    return result.value;
}

/// Checked multiplication with overflow detection via libproven.
pub fn mulChecked(a: i64, b: i64) MathError!i64 {
    const result = c.proven_math_mul_checked(a, b);
    try mapStatus(result.status);
    return result.value;
}

/// Safe absolute value via libproven.
/// Returns error for INT64_MIN (cannot be represented as positive).
pub fn absSafe(n: i64) MathError!i64 {
    const result = c.proven_math_abs_safe(n);
    try mapStatus(result.status);
    return result.value;
}

/// Clamp value to [lo, hi] range via libproven.
pub fn clamp(lo: i64, hi: i64, value: i64) i64 {
    return c.proven_math_clamp(lo, hi, value);
}

/// Integer exponentiation with overflow checking via libproven.
pub fn powChecked(base: i64, exp: u32) MathError!i64 {
    const result = c.proven_math_pow_checked(base, exp);
    try mapStatus(result.status);
    return result.value;
}

test "div" {
    const std = @import("std");
    try std.testing.expectEqual(@as(i64, 5), try div(10, 2));
    try std.testing.expectError(error.DivisionByZero, div(10, 0));
}

test "addChecked" {
    const std = @import("std");
    try std.testing.expectEqual(@as(i64, 3), try addChecked(1, 2));
}

test "clamp" {
    const std = @import("std");
    try std.testing.expectEqual(@as(i64, 5), clamp(0, 10, 5));
    try std.testing.expectEqual(@as(i64, 0), clamp(0, 10, -5));
    try std.testing.expectEqual(@as(i64, 10), clamp(0, 10, 15));
}

File: bindings/rust/src/safe_math.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe arithmetic operations via libproven FFI.
//!
//! All operations delegate to the Idris 2 verified implementation through
//! the Zig FFI layer. No arithmetic logic is implemented in Rust.

use crate::core::{self, Result};
use crate::ffi;

/// Safe math operations that cannot crash.
///
/// All methods call the formally verified Idris 2 implementation via FFI.
pub struct SafeMath;

impl SafeMath {
    /// Safe addition with overflow detection.
    ///
    /// Returns `Err(Error::Overflow)` if the result would exceed `i64::MAX`.
    pub fn add(a: i64, b: i64) -> Result<i64> {
        // SAFETY: proven_math_add_checked is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_add_checked(a, b) };
        core::int_result_to_result(result)
    }

    /// Safe subtraction with underflow detection.
    ///
    /// Returns `Err(Error::Underflow)` if the result would go below `i64::MIN`.
    pub fn sub(a: i64, b: i64) -> Result<i64> {
        // SAFETY: proven_math_sub_checked is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_sub_checked(a, b) };
        core::int_result_to_result(result)
    }

    /// Safe multiplication with overflow detection.
    ///
    /// Returns `Err(Error::Overflow)` if the result would overflow.
    pub fn mul(a: i64, b: i64) -> Result<i64> {
        // SAFETY: proven_math_mul_checked is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_mul_checked(a, b) };
        core::int_result_to_result(result)
    }

    /// Safe division with zero-check.
    ///
    /// Returns `Err(Error::DivisionByZero)` if denominator is 0.
    pub fn div(numerator: i64, denominator: i64) -> Result<i64> {
        // SAFETY: proven_math_div is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_div(numerator, denominator) };
        core::int_result_to_result(result)
    }

    /// Safe modulo with zero-check.
    ///
    /// Returns `Err(Error::DivisionByZero)` if denominator is 0.
    pub fn modulo(numerator: i64, denominator: i64) -> Result<i64> {
        // SAFETY: proven_math_mod is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_mod(numerator, denominator) };
        core::int_result_to_result(result)
    }

    /// Safe absolute value that handles `i64::MIN` correctly.
    ///
    /// Returns `Err(Error::Overflow)` for `i64::MIN` (since its absolute
    /// value cannot be represented as i64).
    pub fn abs(n: i64) -> Result<i64> {
        // SAFETY: proven_math_abs_safe is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_abs_safe(n) };
        core::int_result_to_result(result)
    }

    /// Clamp value to range [lo, hi].
    ///
    /// Always succeeds; returns lo if value < lo, hi if value > hi.
    pub fn clamp(lo: i64, hi: i64, value: i64) -> i64 {
        // SAFETY: proven_math_clamp is a pure function with no
        // pointer arguments; always safe to call.
        unsafe { ffi::proven_math_clamp(lo, hi, value) }
    }

    /// Integer power with overflow checking.
    ///
    /// Returns `Err(Error::Overflow)` if the result would overflow.
    pub fn pow(base: i64, exp: u32) -> Result<i64> {
        // SAFETY: proven_math_pow_checked is a pure function with no
        // pointer arguments; always safe to call.
        let result = unsafe { ffi::proven_math_pow_checked(base, exp) };
        core::int_result_to_result(result)
    }
}
